Implement the shift click gesture to select several adjacent threads in the timeline

julienw opened the pull request:
Gestures that work:
* click some track, shift click some other track => threads in-between are selected. This works in both directions.
* click some track, ctrl click some other track, shift + click a third track => the range is added to the initial selection
* at load time, when there's just one selected track, it's possible to shift + click another track, the existing selected track will be used as the start of the range
* for all of these gestures, after a shift-click, we can shift-click elsewhere, which should cancel the previous shift-click and select again.

Here is a [deploy preview](https://deploy-preview-4138--perf-html.netlify.app/public/gq17n8ck41vfq9vwcbbazd1449dpxr0j6csrtt0/calltree/?globalTrackOrder=0wx8&hiddenGlobalTracks=1wtx0wx6&hiddenLocalTracksByPid=1577709-0wxe~2262903-0w3~1856658-0w3~1577784-0w2~2263414-0w3~1577891-0w5~1642896-0w3~2263332-0w3~1846040-0w3~1776010-0w3~2263283-0w3~2261091-0w3~2262511-0w3~1642608-0w3~2016853-0w3~2024524-0w3~1578256-0wg~2262841-0w3~2020425-0w3~2263091-0w5~2020544-0w3~1579421-0w3~1642983-0w3~1577934-0w5~2262966-0w3~1845149-0w3~1772789-0w3~1577897-0w3~1763983-0w3~1799316-0w5~1578275-0w7~1577831-01~1577849-0w5~1577894-0w6~1577918-0w5~1577956-0w5~1577908-0w8~1578749-0w7~1773756-0w7~1864629-0w278~2263032-0w2568awd&thread=0&timelineType=cpu-category&v=6)

Fixes #2710


https://user-images.githubusercontent.com/454175/180269712-b1c10d66-eee6-44a7-b228-e5caf216f63c.mp4

mstange commented:
This works quite well!

I've noticed one case that I think can be tweaked: "Shift selection after Cmd selection"

 1. Have thread 1 selected.
 2. Cmd+click thread 3.
 3. Shift+click thread 5.

Expected selection: 1, 3, 4, 5 (this is what macOS Finder does)
Actual selection: 3, 4, 5 (everything outside of the new shift selection got discarded)

mstange commented:
I also noticed some weirdness with the IPC tracks, but that's probably something that can be fixed as a follow-up.

For example, in the "Utility AudioDecoder" process, the "MediaPDecoder #1" thread is framed by two IPC tracks. Clicking one of those IPC tracks and then shift-clicking the other creates a selection which only contains the thread-track in between. Cmd+clicking the tracks individually has a different result.

julienw commented:
> This works quite well!
> 
> I've noticed one case that I think can be tweaked: "Shift selection after Cmd selection"
> 
>     1. Have thread 1 selected.
> 
>     2. Cmd+click thread 3.
> 
>     3. Shift+click thread 5.
> 
> 
> Expected selection: 1, 3, 4, 5 (this is what macOS Finder does) Actual selection: 3, 4, 5 (everything outside of the new shift selection got discarded)

Ah, that's interesting!
This works if you keep Ctrl pressed for the shift, but making it work without that would require saving the previous modifier, which sounds more work for this small use case...



> I also noticed some weirdness with the IPC tracks, but that's probably something that can be fixed as a follow-up.
> 
> For example, in the "Utility AudioDecoder" process, the "MediaPDecoder #1" thread is framed by two IPC tracks. Clicking one of those IPC tracks and then shift-clicking the other creates a selection which only contains the thread-track in between. Cmd+clicking the tracks individually has a different result.

Ah that's interesting too! This is quite similar to the previous issue in that when we don't use Ctrl we replace the previous selection. I can see how it's weird in this case though and I'll think about how to fix it easily.

julienw commented:
> > This works quite well!
> > I've noticed one case that I think can be tweaked: "Shift selection after Cmd selection"
> > ```
> > 1. Have thread 1 selected.
> > 
> > 2. Cmd+click thread 3.
> > 
> > 3. Shift+click thread 5.
> > ```
> >     
> > Expected selection: 1, 3, 4, 5 (this is what macOS Finder does) Actual selection: 3, 4, 5 (everything outside of the new shift selection got discarded)
> 
> Ah, that's interesting! This works if you keep Ctrl pressed for the shift, but making it work without that would require saving the previous modifier, which sounds more work for this small use case...
> 

Interestingly: libreoffice spreadsheets works like MacOS X' finder, but google spreadsheets works like my implementation.
Now I'm on the verge of implementing this :)

mstange commented:
Now this case doesn't work as expected anymore:

1 is selected, Cmd 3, Shift 5, Shift 2, expected: 1, 2, 3

I haven't looked at your code at all, but I would implement it as follows:

```js
type State = {|
  currentSelection: Set<TrackId>,
  selectionAtLastNonShiftClick: Set<TrackId>,
  clickedTrackAtLastNoShiftClick: TrackId,
|};

function onClick(state: State, clickedTrack: TrackId): State {
  const newSelection = new Set([clickedTrack]);
  return {
    currentSelection: newSelection,
    selectionAtLastNonShiftClick: newSelection,
    clickedTrackAtLastNoShiftClick: clickedTrack,
  }
}

function onCmdClick(state: State, clickedTrack: TrackId): State {
  const newSelection = new Set([...state.currentSelection, clickedTrack]);
  return {
    currentSelection: newSelection,
    selectionAtLastNonShiftClick: newSelection,
    clickedTrackAtLastNoShiftClick: clickedTrack,
  }
}

function onShiftClick(state: State, clickedTrack: TrackId): State {
  const shiftSelection = computeTracksBetween(state.clickedTrackAtLastNoShiftClick, clickedTrack);
  const newSelection = new Set([...state.selectionAtLastNonShiftClick, ...shiftSelection]);
  return {
    ...state,
    currentSelection: newSelection,
  }
}
```

mstange commented:
> Now this case doesn't work as expected anymore:

Oh, correction, it wasn't working at all before either. So I should rather say: Thank you for implementing the Cmd-then-shift behavior, but here's a case that can still be improved to match macOS Finder more closely.

julienw commented:
> Now this case doesn't work as expected anymore:
> 
> 1 is selected, Cmd 3, Shift 5, Shift 2, expected: 1, 2, 3

I agree, I expect this too work too, and it doesn't. This works as expected without the Ctrl thing.

> 
> I haven't looked at your code at all, but I would implement it as follows:
> 
> ```js
> type State = {|
>   currentSelection: Set<TrackId>,
>   selectionAtLastNonShiftClick: Set<TrackId>,
>   clickedTrackAtLastNonShiftClick: TrackId,
> |};
> 
> function onClick(state: State, clickedTrack: TrackId): State {
>   const newSelection = new Set([clickedTrack]);
>   return {
>     currentSelection: newSelection,
>     selectionAtLastNonShiftClick: newSelection,
>     clickedTrackAtLastNonShiftClick: clickedTrack,
>   }
> }
> 
> function onCmdClick(state: State, clickedTrack: TrackId): State {
>   const newSelection = new Set([...state.currentSelection, clickedTrack]);
>   return {
>     currentSelection: newSelection,
>     selectionAtLastNonShiftClick: newSelection,
>     clickedTrackAtLastNonShiftClick: clickedTrack,
>   }
> }
> 
> function onShiftClick(state: State, clickedTrack: TrackId): State {
>   const shiftSelection = computeTracksBetween(state.clickedTrackAtLastNonShiftClick, clickedTrack);
>   const newSelection = new Set([...state.selectionAtLastNonShiftClick, ...shiftSelection]);
>   return {
>     ...state,
>     currentSelection: newSelection,
>   }
> }
> ```

This is pretty close to what I did but maybe I added piles on top of other piles and I could now possibly simplify this a bit. For example my code used to work in a quite stateless way initially, always recomputing the full range when shift was used and replacing the full selection, but then I added more state to control it. I like how you wrote that shift click reuse the previous selection, which builds on the assumption that the previous click either already replaced the previous selection or added to the previous selection, and I haven't thought of this idea before. So thanks!

julienw commented:
OK, I know what my bug is, and what you propose (saving the `selectionAtLastNonShiftClick`) is indeed the right way to implement this. My code is always adding selections during this STR, while we want to replace what was added instead.
Thanks for the feedback!

julienw commented on src/components/timeline/TrackThread.js:
```diff
@@ -125,45 +124,39 @@ class TimelineTrackThreadImpl extends PureComponent<Props> {
-          invertCallstack,
-          selectedThreadIndexes,
-        } = this.props;
+    const modifiers = getTrackSelectionModifiers(event);
```
Here it looks like it's a lot of change, but if you look at the changes without the whitespace changes you'll see the code has just been reordered.

julienw commented on src/actions/profile-view.js:
```diff
@@ -269,6 +269,122 @@ export function changeSelectedThreads(
+ * information in the state. Because of all possible cases this isn't trivial.
+ * This will return null for tracks that are not selectable.
+ */
+function getInformationFromTrackReference(
```
This commit might be easier to follow _with_ whitespace changes.

julienw commented:
I reimplemented most of the patch following Markus' excellent suggestion (I wish I thought of that earlier :D), now I believe all mentioned use cases work as expected.

It should be easier to look commit by commit.

julienw commented on src/actions/profile-view.js:
```diff
@@ -269,154 +276,436 @@ export function changeSelectedThreads(
+      return;
+    }
+
+    let selectedTab =
+      clickedTrackInformation.relatedTab ?? getSelectedTab(getState());
+    const visibleTabs = getThreadSelectors(
+      clickedTrackInformation.relatedThreadIndex
+    ).getUsefulTabs(getState());
     if (!visibleTabs.includes(selectedTab)) {
       // If the user switches to another track that doesn't have the current
       // selectedTab then switch to the first tab.
       selectedTab = visibleTabs[0];
     }
```
I wasn't sure what to do with the `selectedTab`, should we change it for all modes? Maybe we'd want to change it only for `setOneTrackSelection`... I thought it would be easy to change in a follow-up if needed.

canova commented on src/actions/profile-view.js:
```diff
@@ -446,9 +467,158 @@ function toggleOneTrack(
+ * This computes the set of threads that's between two tracks (including
+ * themselves). This skips over hidden tracks.
+ */
+function findThreadsBetweenTracks(
```
Even tough this function is pretty big, it's very easy to follow. Thanks!

canova commented on src/actions/profile-view.js:
```diff
@@ -446,9 +467,158 @@ function toggleOneTrack(
+      }
+    }
+
+    // Add the global track if it's not a virtual track and not out of the range.
```
By "virtual" you mean a global tack with a thread, right?

julienw replied:
"virtual track" means here a global track that is _not_ a thread (such as what we get when importing sometimes).

But the comment says "add the global track if it's _not_ virtual", maybe that's where you got confused?

canova commented on src/actions/profile-view.js:
```diff
@@ -609,10 +615,18 @@ function selectRangeOfTracks(
+      const selectedThreadIndexes = getSelectedThreadIndexes(getState());
+      const threadIndex = getFirstItemFromSet(selectedThreadIndexes);
+      if (threadIndex !== undefined) {
+        lastClickedTrack = getTrackReferenceFromThreadIndex(
```
It's a bit unfortunate that we had to do a lot of track reference <-> thread index conversions. But it's not possible to fix it in this PR.

julienw replied:
yeah, it's not obvious how we could change this too...

canova reviewed (approved):
Looks great, thanks for the feature and adding lots of tests! I've added a few nits mostly.

julienw commented on src/utils/index.js:
```diff
@@ -96,22 +96,16 @@ export function getStartEndRangeForMarker(
-  //   return 'shift';
-  // }
-  return 'none';
+): {| ctrlOrMeta: boolean, shift: boolean |} {
```
yeah, I thought of that. I wasn't quite sure where to put it :D maybe types/actions... it's not used in an action payload but it's used in an action function parameter.

julienw commented on src/actions/profile-view.js:
```diff
@@ -272,11 +273,16 @@ export function changeSelectedThreads(
+  // This is the track index for the global track where this track is located.
+  globalTrackIndex: TrackIndex,
+  // This is the track index of the local track in its process group.
+  localTrackIndex: null | TrackIndex,
```
yeah, good idea.
I was also thinking that the  property `type: 'global' | 'local'` could be infered from that, but thought that being more explicit would make the code easier to follow (possibly).

julienw commented on src/actions/profile-view.js:
```diff
@@ -280,6 +280,8 @@ type TrackInformation = {|
   // This is the tab that should be selected from this track. `null` if this
   // track doesn't have a prefered tab.
   relatedTab: null | TabSlug,
+  // This is the initial track reference
```
yeah exactly, I'll improve the comment.
I need this so that we can save it to the state later. I was also thinking to store a `TrackInformation` instead, but thought that `TrackInformation` is more a "temporary data exchange" structure rather than something to be stored...

julienw commented:
Thanks so much for the review! I believe I handled all your comments and will merge now!
